### ISSUE TITLE
[develop] Stop supporting themes that require prompinit and its prompt_name_setup file to be on the fpath

### ISSUE THREAD

**Issue by ericbn** [zimfw/zimfw#325]
While implementing the develop branch of Zim with the new "plugin mechanism" (see #323) I could not come up with a nice and simple implementation to support themes that use promptinit and have prompt_name_setup files that need to be added to the fpath.

Then I though about it and first realized that promptinit does not seem practical or useful for the everyday usage. This is what it offers:
* Autoloading themes with `prompt_*_setup` files in the `fpath`, so you can have a list of all the themes "installed" without actually running them.
* Listing, previewing, getting help for, and switching between the themes "installed" with the `prompt` function.
* Allowing themes to set up the prompt Zsh options with the `prompt_opts` array instead of using `setopt` directly. This is a **bad one**, because it makes themes absolutely dependent on `prompt`. Zsh <5.4 had a bug where `sp` was not taken into account, and this broke themes starting at Zsh 5.4 that didn't have it (promptsp is a default option and it would be unset then).
* Cleaning up before previewing or mainly before switching to another theme with `prompt`. Early versions cleaned up `zle_highlight[(r)default:*]`. With Zsh 4.3.17 it also cleaned up hooks. (Not even used the `prompt_opts` array to clean the options up!) Starting with Zsh 5.4, `prompt_cleanup` was introduced to define a custom cleanup command; and PS* variables, precmd and preexec hooks, and prompt_opts defined by the theme are cleaned up by default.

Look at all those items above, and think what's the use of them if you just use a single prompt all through your shell sessions. None, right? Now think how many times in the last year have you experimented with lots of prompts (if you didn't start using Zsh in the last year of course). And must be lots, so "brute-force" starting a new shell to experiment each new one would be a burden! None again, right?

Next thing I thought was: let me check how many themes out there actually require `promptinit`. Because it's the Wild West out there! I know our themes do. And prezto themes too. I know also oh-my-zsh themes don't. I'll show my finding about this in a comment below.

**Comment by ericbn** [comment]
@robobenklein, the way I see that themes should best behave is:

* If they're sourced directly (calling `promptinit` after that has no effect, so not needed), then they should be activated and no cleaning up process would be set up. The idea is that when you're sourcing a theme directly you mean to use it for the whole shell session.
* If their `prompt_name_setup` file is added to the `fpath` before `promptinit` is called, then they can be listed, previewed, and show their help with `prompt`. Also they should be activated and cleaned up with `prompt`.

To support both behaviors above, a theme should have:

* A `prompt_name_setup` file that defines a `prompt_name_setup` function. The file should call the function when sourced, and the function should activate the theme. The function can/should set the following:
  * `PS1`, `RPS1` and other prompt variables. Should use these short names instead of `PROMPT`, `RPROMPT`, etc.
  * Optionally call `add-zsh-hook` to add `prompt_name_{chpwd,precmd,preexec,periodic,zshaddhistory,zshexit}` hooks.
  * Optionally set `zle_highlight[(r)default:*]`.
* The file (which can be sourced or serve as an autoloaded function itself) can also define the following:
  * Any hook functions that are added during the `prompt_name_setup` function (see above).
  * An optional `prompt_name_help` function that prints out the theme help message.
  * An optional `prompt_name_preview` function that shows custom theme previews.

These are things that work with `prompt`, but don't work when sourcing the file directly, so, to support both behaviors, these **should not be used**:
* Not autoloading `add-zsh-hook` (sorry for the double negative) (`prompt` does that for you, but the function will not be already autoloaded if you source the theme directly).
* Using the `prompt_opts` array instead of calling `setopt`. This has the drawback of not allowing `prompt` to properly cleanup the options set by the theme, and I think this should be improved on the `prompt` function side (it should also clean up prompt options set with `setopt`). @bartschaefer, what do you think? See an analysis about this in the issue description and comments above.

And this is something that can be optionally used: Allowing the `prompt_name_setup` function (and the file / autoloaded function that calls it) to take parameters.

* Usually plugin mechanisms don't allow passing parameters to sourced scripts (we plan to keep it that way with Zim too). So when sourced directly, the theme should be able to be configured with environment variables or with zstyle. Themes should always allow configuration with environment variables or zstyle (when they have any configuration of course).
* If you allow the theme to also be configured with parameters, then I suggest you make the function take the parameters passed and set the correspondent environment variables or zstyle styles. Allowing configuration with parameters is optional, and a nice to have if you want to be friendly with `prompt`.

**Comment by PatTheMav** [comment]
After all my work to support `promptinit` in https://github.com/PatTheMav/minimal2? 😱

But should be easy enough to migrate to `setopt` calls and move the function parts around, will take a look at it later tonight.

**Comment by ericbn** [comment]
@PatTheMav, just pushed changes to our themes for the new version, following the recommendations I make here. See commits referenced above. I see you just [removed the prompt_opts dependency](https://github.com/PatTheMav/minimal2/commit/b459e18c059713ffedcbadb62fb4e383ffb51b6d). You can also delete the functions/prompt_minimal2_setup file.

**Comment by PatTheMav** [comment]
@Eriner Yeah got lucky with that theme I guess, as it followed many requirements already. For some reason it's not properly loaded within the current develop branch of zimfw. I always have to manually `source` the setup file to make it work... 🤔

**Comment by ericbn** [comment]
@PatTheMav, please do `zimfw upgrade`. I just pushed the necessary changes that will close this ticket.

**Comment by PatTheMav** [comment]
@ericbn Done, the prompt expansion works as intended again. However I'm always greeted by the `prompt` help screen when starting a new terminal session:

```
Last login: Sun Jan 20 01:46:00 on ttys000
Usage: prompt <options>
Options:
    -c              Show currently selected theme and parameters
    -l              List currently available prompt themes
    -p [<themes>]   Preview given themes (defaults to all)
    -h [<theme>]    Display help (for given theme)
    -s <theme>      Set and save theme
    <theme>         Switch to new theme immediately (changes not saved)

Use prompt -h <theme> for help on specific themes.
  ›                                                                                                                                                          ~
```

**Comment by ericbn** [comment]
@PatTheMav, you can remove `prompt` from your modules list now. Look at the latest changes in the zimrc template here: https://github.com/zimfw/zimfw/commit/dca2cf111e22f1ef9c0859f5e2b15d834af91627#diff-c462b17081d99f90bf394768c03a4415

**Comment by PatTheMav** [comment]
@ericbn Yeah, works as intended now. Still wonder if the theme is still compatible with `promptinit` now, but that doesn't concern this issue.

**Comment by ericbn** [comment]
Yeah, looks like it's not. Same happens with our Zim themes. I missed one thing in my analysis of the promptinit code: it will clear all prompt options if prompt_opts is empty in Zsh 5.4, but not in earlier versions. My bad, I was testing with 5.3. Another reason to run from promptinit like the plague  :- )

I'll change my recommendations above. Code should do both now:

    # Support promptinit, but don't depend on it, so also do setopt
    setopt nopromptbang promptcr promptpercent promptsp promptsubst
    prompt_opts=(cr percent sp subst)

Thanks for you collaboration with the develop version of Zim!

I've invited you to move your repo to the zimfw organization, since that's the idea going forward when the develop branch gets promoted, and it will be great to have you as the first collaborator to have its own repo here. (@Eriner, hope you're good with this. I believe that's one of the many ideas we have in common for the project. Right?)

**Comment by ericbn** [comment]
@robobenklein, my recommendations for a theme that supports promptinit but does not depend on it have changed. My initial recommendations were only correct for Zsh <5.4. I've updated [them here](https://github.com/zimfw/zimfw/issues/325#issuecomment-455426700), but the TLDR; of the update is that you should "**keep using prompt_opts** and also make sure you set/unset the same `prompt{bang,cr,percent,sp,subst}` options with `setopt`". Hopefully that will change in future versions of Zsh and everyone will forget prompt_opts ever existed!  :- )

**Comment by ericbn** [comment]
**TL;DR** Just forget all the obsessive details and going back and forth in the comments above. That was `promptinit` driving me crazy. Just forget it exists, don't support it at all, and enjoy your life!

**Comment by PatTheMav** [comment]
FYI, I received a bug report on my own theme repo about weird behavior of the theme with `zimfw`'s master branch and the tl;dr is:

- Theme was loaded automatically
- Promptinit was still called
- Weird behavior ensues

Apart from telling people to disable the `prompt` module entirely I haven't found a "nice" way to fix this, but will probably just update the readme file.
